src/storage/src/source/kafka.rs

```diff
@@ -634,140 +518,200 @@ fn render_reader<G: Scope<Timestamp = KafkaTimestamp>>(
+            let mut metadata_update: Option<MetadataUpdate> = None;
             let mut snapshot_total = None;
 
             let max_wait_time =
                 mz_storage_types::dyncfgs::KAFKA_POLL_MAX_WAIT.get(config.config.config_set());
             loop {
-                let partition_info = reader.partition_info.lock().unwrap().take();
-                if let Some((probe_ts, partitions)) = partition_info {
-                    let max_pid = partitions.keys().last().cloned();
-                    let lower = max_pid
-                        .map(RangeBound::after)
-                        .unwrap_or(RangeBound::NegInfinity);
-                    let future_ts =
-                        Partitioned::new_range(lower, RangeBound::PosInfinity, MzOffset::from(0));
-
-                    // Topics are identified by name but it's possible that a user recreates a
-                    // topic with the same name but different configuration. Ideally we'd want to
-                    // catch all of these cases and immediately error out the source, since the
-                    // data is effectively gone. Unfortunately this is not possible without
-                    // something like KIP-516 so we're left with heuristics.
-                    //
-                    // The first heuristic is whether the reported number of partitions went down
-                    if !PartialOrder::less_equal(data_cap.time(), &future_ts) {
-                        let prev_pid_count = prev_pid_info.map(|info| info.len()).unwrap_or(0);
-                        let pid_count = partitions.len();
-                        let err = DataflowError::SourceError(Box::new(SourceError {
-                            error: SourceErrorDetails::Other(
-                                format!(
-                                    "topic was recreated: partition \
-                                     count regressed from {prev_pid_count} to {pid_count}"
-                                )
-                                .into(),
-                            ),
-                        }));
-                        let time = data_cap.time().clone();
-                        let err = Err(err);
-                        for (output, err) in
-                            outputs.iter().map(|o| o.output_index).repeat_clone(err)
-                        {
-                            data_output
-                                .give_fueled(&data_cap, ((output, err), time, 1))
-                                .await;
+                // Wait for data or metadata events while also making progress with offset
+                // committing.
+                tokio::select! {
+                    // TODO(petrosagg): remove the timeout and rely purely on librdkafka waking us
+                    // up
+                    _  = tokio::time::timeout(max_wait_time, notificator.notified()) => {},
+
+                    Some(Event::Data(_, mut updates)) = metadata_input.next() => {
```

Comment:
This will process at most one piece of metadata per loop, but that might lead to a bunch of metadata probes being accumulated in the buffers. I think we should select on `metadata_input.ready()` and in the body run a loop over `metadata_input.next_sync()` to exhaust the available input

Reply:
Makes sense, done!